src/mqtt_machine.erl

```diff
@@ -0,0 +1,121 @@
+%% The contents of this file are subject to the Mozilla Public License
+%% Version 1.1 (the "License"); you may not use this file except in
+%% compliance with the License. You may obtain a copy of the License
+%% at https://www.mozilla.org/MPL/
+%%
+%% Software distributed under the License is distributed on an "AS IS"
+%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+%% the License for the specific language governing rights and
+%% limitations under the License.
+%%
+%% The Original Code is RabbitMQ.
+%%
+%% The Initial Developer of the Original Code is GoPivotal, Inc.
+%% Copyright (c) 2007-2019 Pivotal Software, Inc.  All rights reserved.
+%%
+-module(mqtt_machine).
+-behaviour(ra_machine).
+
+-include_lib("ra/include/ra.hrl").
+
+-export([init/1,
+         apply/3,
+         state_enter/2]).
+
+-record(state, {client_ids = #{}}).
+
+-type state() :: #state{}.
+
+-type config() :: map().
+
+-type reply() :: {ok, term()} | {error, term()}.
+-type client_id() :: term().
+
+-type command() :: {register, client_id(), pid()} |
+                   {unregister, client_id(), pid()} |
+                   list.
+
+-spec init(config()) -> state().
+init(_Conf) ->
+    #state{}.
+
+-spec apply(map(), command(), state()) ->
+    {state(), reply(), ra_machine:effects()}.
+apply(Meta, {register, ClientId, Pid}, #state{client_ids = Ids} = State0) ->
+    {Effects, Ids1} =
+        case maps:find(ClientId, Ids) of
+            {ok, OldPid} when Pid =/= OldPid ->
+                Effects0 = [{demonitor, process, OldPid},
+                            {monitor, process, Pid},
+                            {mod_call, gen_server2, cast, [OldPid, duplicate_id]}],
+                {Effects0, maps:remove(ClientId, Ids)};
+            error ->
+              Effects0 = [{monitor, process, Pid}],
+              {Effects0, Ids}
+        end,
+    State = State0#state{client_ids = maps:put(ClientId, Pid, Ids1)},
+    {State, ok, Effects ++ snapshot_effects(Meta, State)};
+
+apply(Meta, {unregister, ClientId, Pid}, #state{client_ids = Ids} = State0) ->
+    State = case maps:find(ClientId, Ids) of
+      {ok, Pid}         -> State0#state{client_ids = maps:remove(ClientId, Ids)};
+      %% don't delete client id that might belong to a newer connection
+      %% that kicked the one with Pid out
+      {ok, _AnotherPid} -> State0;
+      error             -> State0
+    end,
+    {State, ok, [{demonitor, process, Pid}] ++ snapshot_effects(Meta, State)};
```

Comment:
Issuing a snapshot effect on every command would work as Ra itself won't actually take a snapshot unless at least 4096 (I think) commands have been processed since the last one. For this kind of state machine where we snapshot the entire state but the state is likely to be fairly small it is probably fine. The ideal would be to estimate the size of the log vs the size of the state machine and only emit release cursors when it would be beneficial to snapshot. E.g. after keys have been deleted.

Reply:
I'm not really sure what the suggestion is here. I made the snapshot effect conditional, emitted only when the map actually changes. Without introducing operation rate tracking my best idea is to emit periodically or speculatively (e.g. with a probability of roughly N).